Python/ceval.c

```diff
@@ -439,35 +439,83 @@ int pthread_attr_destroy(pthread_attr_t *a)
 #endif
 
 
+static void
+tstate_set_stack(PyThreadState *tstate,
+                 void *stack_start_addr, size_t stack_size)
+{
+    assert(stack_size > 0);
+
+    _PyThreadStateImpl *ts = (_PyThreadStateImpl *)tstate;
+    ts->c_stack_hard_limit = (uintptr_t)stack_start_addr;
+    ts->c_stack_top = (uintptr_t)stack_start_addr + stack_size;
+
+    uintptr_t soft_limit = ts->c_stack_hard_limit;
+    if (stack_size >= _PyOS_STACK_MARGIN_BYTES) {
+#ifdef _Py_THREAD_SANITIZER
+        // Thread sanitizer crashes if we use a bit more than half the stack.
+        soft_limit += (stack_size / 2);
+#else
+        soft_limit += _PyOS_STACK_MARGIN_BYTES;
+#endif
+    }
+    ts->c_stack_soft_limit = soft_limit;
+
+    // Sanity checks
+    assert(ts->c_stack_hard_limit <= ts->c_stack_soft_limit);
+    assert(ts->c_stack_soft_limit < ts->c_stack_top);
+
+    // Test the stack pointer
+#if !defined(NDEBUG) && !defined(__wasi__)
```

Comment:
These assertions fail on WASI but I don't have time right now to investigate why.

Reply:
So, unless we specifically say that `PyUnstable_ThreadState_SetStack()` should only be called after the stack switch, we'll need to remove these asserts.

Reply:
Oh you're right. I moved the assertions on the stack pointer to `PyUnstable_ThreadState_ResetStack()`.